drv/lpc55-swd/src/main.rs

```diff
@@ -512,28 +598,130 @@ impl idl::InOrderSpCtrlImpl for ServerImpl {
+            + notifications::SP_RESET_IRQ_MASK
+            + notifications::JTAG_DETECT_IRQ_MASK
     }
 
-    fn handle_notification(&mut self, _bits: u32) {
-        ringbuf_entry!(Trace::WatchdogFired);
+    fn handle_notification(&mut self, bits: u32) {
+        // If JTAG_DETECT fires:
+        //   - invalidate any SP measurement
+        //   - if still asserted, then the other handlers will fail on
+        //     their calls to do_setup_swd();
+        //   - We could try extending the Watchdog timer if it is active in hopes
+        //     that the SP dongle is removed or its power is removed, but if
+        //     there is a dongle attached to the SP, then let the humans figure it out
+        //     and don't complicate the behavior here.
+        //
 
-        // Disable the watchdog timer
-        sys_set_timer(None, notifications::TIMER_MASK);
+        let mut invalidate = false;
+        let gpio = Pins::from(self.gpio);
+
+        if (bits & notifications::JTAG_DETECT_IRQ_MASK) != 0 {
+            ringbuf_entry!(Trace::SpJtagDetectFired);
+            const SLOT: PintSlot = SP_TO_ROT_JTAG_DETECT_L_PINT_SLOT;
+            if gpio
+                .pint_op(SLOT, PintOp::Detected, PintCondition::Falling)
+                .map_or(false, |v| v.unwrap_lite())
+            {
+                ringbuf_entry!(Trace::InvalidateSpMeasurement);
+                invalidate = true;
+                self.init = false;
+            }
+            let _ = gpio.pint_op(SLOT, PintOp::Clear, PintCondition::Status);
+            sys_irq_control(notifications::JTAG_DETECT_IRQ_MASK, true);
+        }
 
-        // Attempt to do the swap
-        let r = self.swap_sp_slot();
-        ringbuf_entry!(Trace::WatchdogSwap(r));
+        if (bits & notifications::TIMER_MASK) != 0 {
+            ringbuf_entry!(Trace::WatchdogFired);
 
-        // Force reinitialization
-        self.init = false;
+            match self.do_setup_swd() {
+                Ok(()) => {
+                    // Disable the watchdog timer
+                    sys_set_timer(None, notifications::TIMER_MASK);
+                    // Attempt to do the swap
+                    let r = self.swap_sp_slot();
+                    ringbuf_entry!(Trace::WatchdogSwap(r));
+
+                    // Force reinitialization
+                    self.init = false;
+                }
+                Err(e) => {
+                    // This should only fail if JTAG_DETECT or SP_RESET are currently asserted.
+                    ringbuf_entry!(Trace::TimerHandlerError(e));
+                }
+            }
+            self.watchdog_ms = None;
+        }
+
+        if (bits & notifications::SP_RESET_IRQ_MASK) != 0 {
+            ringbuf_entry!(Trace::SpResetFired);
+            if !invalidate && !self.do_handle_sp_reset() {
+                ringbuf_entry!(Trace::InvalidateSpMeasurement);
+                invalidate = true;
+            }
+            // else something something {
```

Comment:
Remains of an else state here? (see also `//}` at L708)

Reply:
gone